ts/utils/web3_connector.ts

```diff
@@ -0,0 +1,224 @@
+        const provider = this._contractWrappers.getProvider();
+        const web3Wrapper = new Web3Wrapper(provider);
+        const userAccountsIfExists = await web3Wrapper.getAvailableAddressesAsync();
+        return userAccountsIfExists;
+    }
+    public async getZrxBalanceAsync(owner: string): Promise<BigNumber> {
+        utils.assert(this._contractWrappers !== undefined, 'ContractWrappers must be instantiated.');
+        const contractAddresses = getContractAddressesForNetworkOrThrow(this.networkId);
+        const tokenAddress: string = contractAddresses.zrxToken;
+        const erc20Token = new ERC20TokenContract(tokenAddress, this._contractWrappers.getProvider());
+        try {
+            const amount = await erc20Token.balanceOf.callAsync(owner);
+            return amount;
+        } catch (error) {
+            return ZERO;
+        }
+    }
+
+    public async connectToWalletAsync(provider: Providers): Promise<ConnectedWalletDetails | undefined> {
+        const networkIdIfExists = await this._getInjectedProviderNetworkIdIfExistsAsync();
+        this.networkId = networkIdIfExists !== undefined ? networkIdIfExists : constants.NETWORK_ID_MAINNET;
+
+        await this._resetOrInitializeAsync(this.networkId);
+
+        await this._fetchAddressesAndBalancesAsync();
+
+        // Always assume selected index is 0 for Metamask
+        const accountInfo = await this._updateSelectedAddressAsync(0);
+        const providerName = providerToName[provider] || constants.PROVIDER_NAME_GENERIC;
+
+        return {
+            ...accountInfo,
+            provider,
+            providerName,
+        };
+    }
+
+    private async _fetchAddressesAndBalancesAsync(): Promise<void> {
+        const userAddresses = await this._getUserAddressesAsync();
+        const addressBalances: BigNumber[] = [];
+        for (const address of userAddresses) {
+            const balanceInZrx = await this.getZrxBalanceAsync(address);
+            addressBalances.push(balanceInZrx);
+        }
+
+        this.userAddresses = userAddresses;
+        this.addressBalances = addressBalances;
+    }
+
+    private async _updateSelectedAddressAsync(
+        index: number,
+    ): Promise<{ selectedAddress: string; currentBalance: BigNumber }> {
+        const { userAddresses, addressBalances } = this;
+        if (!userAddresses[index]) {
+            throw new Error(`Could not get address at index ${index}`);
+        }
+
+        return {
+            selectedAddress: userAddresses[index],
+            currentBalance: addressBalances[index],
+        };
+    }
+
+    private _getNameGivenProvider(provider: ZeroExProvider): string {
+        const providerType = utils.getProviderType(provider);
+        const providerNameIfExists = providerToName[providerType];
+        if (providerNameIfExists === undefined) {
+            return constants.PROVIDER_NAME_GENERIC;
+        }
+        return providerNameIfExists;
+    }
+    private async _getProviderAsync(
```

Comment:
This code might be helpful. https://github.com/0xProject/0x-monorepo/blob/development/packages/instant/src/util/provider_factory.ts 

as well as this code.

https://github.com/0xProject/0x-monorepo/blob/development/packages/instant/src/util/provider_state_factory.ts

Potential alternative architecture:
I would love to have a similar structure to Instant. Ie. we store a lot of the web3 info in the redux state, and do not rely on initializing this class to get access to it.
When we do need to make an RPC call, I think it would be best to lean on web3Wrapper (which would be in the redux state) and pass it into a utility function.

Reply:
I really like this approach! I did not realize this this existed 😳 

I will refactor to use these factories. Longer term it might even make sense to move this into module that can be shared between instant and the website. 👍